<a href="https://colab.research.google.com/github/edsrisk/portfolio_optimization/blob/main/dow_jones_optimal_portfolio.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Portfolio Optimization: Creating an optimal portfolio of Dow Jones stocks for period 1/4/2021 to 7/23/2021

credit to: Computer Science youtube channel

In [41]:
# Import libraries and dependencies
import pandas as pd
import numpy as np
import requests
import matplotlib.pyplot as plt

In [42]:
# Load NYSE data
from google.colab import files
files.upload()

{}

In [44]:
# read source file
dow_opt = pd.read_csv('dow_optimization.csv')
dow_opt.head()

Date        AAPL        AMGN  ...        WBA         WMT         VZ
0  1/4/2021  128.997803  223.416199  ...  40.658897  145.349930  56.966927
1  1/5/2021  130.592697  224.500443  ...  40.423191  144.576218  56.705570
2  1/6/2021  126.196747  229.911880  ...  42.259716  145.478897  57.392853
3  1/7/2021  130.502991  230.670868  ...  44.449795  145.468964  57.263702
4  1/8/2021  131.629379  235.076889  ...  44.400692  145.449127  56.549496

[5 rows x 31 columns]

In [45]:
# Do a quick inspection of the file
dow_opt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140 entries, 0 to 139
Data columns (total 31 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Date    140 non-null    object 
 1   AAPL    140 non-null    float64
 2   AMGN    140 non-null    float64
 3   AXP     140 non-null    float64
 4   BA      140 non-null    float64
 5   CAT     140 non-null    float64
 6   CRM     140 non-null    float64
 7   CSCO    140 non-null    float64
 8   CVX     140 non-null    float64
 9   DIS     140 non-null    float64
 10  DOW     140 non-null    float64
 11  GS      140 non-null    float64
 12  HD      140 non-null    float64
 13  HON     140 non-null    float64
 14  IBM     140 non-null    float64
 15  INTC    140 non-null    float64
 16  JNJ     140 non-null    float64
 17  JPM     140 non-null    float64
 18  KO      140 non-null    float64
 19  MCD     140 non-null    float64
 20  MMM     140 non-null    float64
 21  MRK     140 non-null    float64
 22  MS

In [46]:
# Index on Data
dow_opt = dow_opt.set_index(pd.DatetimeIndex(dow_opt['Date'].values))
dow_opt.head()

Date        AAPL        AMGN  ...        WBA         WMT         VZ
2021-01-04  1/4/2021  128.997803  223.416199  ...  40.658897  145.349930  56.966927
2021-01-05  1/5/2021  130.592697  224.500443  ...  40.423191  144.576218  56.705570
2021-01-06  1/6/2021  126.196747  229.911880  ...  42.259716  145.478897  57.392853
2021-01-07  1/7/2021  130.502991  230.670868  ...  44.449795  145.468964  57.263702
2021-01-08  1/8/2021  131.629379  235.076889  ...  44.400692  145.449127  56.549496

[5 rows x 31 columns]

In [47]:
dow_opt.info()
# Note the change in date format

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 140 entries, 2021-01-04 to 2021-07-23
Data columns (total 31 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Date    140 non-null    object 
 1   AAPL    140 non-null    float64
 2   AMGN    140 non-null    float64
 3   AXP     140 non-null    float64
 4   BA      140 non-null    float64
 5   CAT     140 non-null    float64
 6   CRM     140 non-null    float64
 7   CSCO    140 non-null    float64
 8   CVX     140 non-null    float64
 9   DIS     140 non-null    float64
 10  DOW     140 non-null    float64
 11  GS      140 non-null    float64
 12  HD      140 non-null    float64
 13  HON     140 non-null    float64
 14  IBM     140 non-null    float64
 15  INTC    140 non-null    float64
 16  JNJ     140 non-null    float64
 17  JPM     140 non-null    float64
 18  KO      140 non-null    float64
 19  MCD     140 non-null    float64
 20  MMM     140 non-null    float64
 21  MRK     140 non-null

In [48]:
dow_opt.columns

Index(['Date', 'AAPL', 'AMGN', 'AXP', 'BA', 'CAT', 'CRM', 'CSCO', 'CVX', 'DIS',
       'DOW', 'GS', 'HD', 'HON', 'IBM', 'INTC', 'JNJ', 'JPM', 'KO', 'MCD',
       'MMM', 'MRK', 'MSFT', 'NKE', 'PG', 'TRV', 'UNH', 'V', 'WBA', 'WMT',
       'VZ'],
      dtype='object')

In [49]:
# Drop the date column
dow_opt.drop(columns=['Date'],axis=1,inplace=True)

In [50]:
dow_opt.head()

AAPL        AMGN  ...         WMT         VZ
2021-01-04  128.997803  223.416199  ...  145.349930  56.966927
2021-01-05  130.592697  224.500443  ...  144.576218  56.705570
2021-01-06  126.196747  229.911880  ...  145.478897  57.392853
2021-01-07  130.502991  230.670868  ...  145.468964  57.263702
2021-01-08  131.629379  235.076889  ...  145.449127  56.549496

[5 rows x 30 columns]

Portfolio Analysis: Current State

In [51]:
# Calculate daily returns
dow_opt_ret = dow_opt.pct_change()
dow_opt_ret.head()

AAPL      AMGN       AXP  ...       WBA       WMT        VZ
2021-01-04       NaN       NaN       NaN  ...       NaN       NaN       NaN
2021-01-05  0.012364  0.004853  0.005337  ... -0.005797 -0.005323 -0.004588
2021-01-06 -0.033662  0.024104  0.036993  ...  0.045432  0.006244  0.012120
2021-01-07  0.034123  0.003301 -0.007910  ...  0.051824 -0.000068 -0.002250
2021-01-08  0.008631  0.019101  0.000986  ... -0.001105 -0.000136 -0.012472

[5 rows x 30 columns]

In [52]:
# Calculate the annualized covariance of returns across all 30 dow stocks
dow_opt_cov = dow_opt_ret.cov()*252
dow_opt_cov

AAPL      AMGN       AXP  ...       WBA       WMT        VZ
AAPL  0.073677  0.013981  0.002776  ...  0.004019  0.014040 -0.002546
AMGN  0.013981  0.039333  0.000941  ...  0.010603  0.009843  0.008429
AXP   0.002776  0.000941  0.073079  ...  0.008663  0.005663  0.010125
BA    0.020996  0.003895  0.049892  ...  0.025226  0.018325  0.004702
CAT  -0.005445  0.006989  0.043835  ...  0.015737  0.008367  0.009271
CRM   0.035257  0.004365  0.006093  ... -0.005342  0.010721  0.000024
CSCO  0.021290  0.013673  0.013930  ...  0.016363  0.011023  0.007329
CVX   0.000904  0.005659  0.046608  ...  0.017432  0.007809  0.013288
DIS   0.004293  0.005482  0.040374  ...  0.006185  0.008110  0.010189
DOW   0.007850  0.009607  0.052985  ...  0.021893  0.015087  0.012989
GS    0.003171  0.000664  0.043679  ...  0.023559  0.010186  0.009224
HD    0.017033  0.012471  0.009741  ...  0.010965  0.014285  0.011178
HON   0.010945  0.006976  0.034296  ...  0.014119  0.009395  0.010410
IBM   0.000591  0.009278  0.017445  ...  0.017488  0.006958  0.007864
INTC  0.039006  0.012030  0.027827  ...  0.010407  0.008886  0.001649
JNJ   0.006483  0.014212  0.004499  ...  0.010864  0.009235  0.007647
JPM   0.000038  0.002652  0.038398  ...  0.017712  0.006578  0.011171
KO    0.010740  0.005918  0.009815  ...  0.002522  0.008356  0.007723
MCD   0.014671  0.010658  0.010987  ...  0.007990  0.009560  0.007209
MMM   0.005742  0.005277  0.013682  ...  0.019016  0.005135  0.006208
MRK  -0.001920  0.018015  0.002514  ...  0.011924  0.003607  0.009374
MSFT  0.041884  0.013368  0.000565  ...  0.004202  0.010502 -0.000758
NKE   0.019101  0.011110  0.018704  ...  0.004953  0.010435  0.005832
PG    0.003969  0.010609  0.002966  ...  0.004651  0.010509  0.011758
TRV   0.009075  0.011366  0.036536  ...  0.009755  0.010330  0.016426
UNH   0.002244  0.011935  0.013284  ...  0.003778  0.007150  0.010664
V     0.023283  0.005825  0.025426  ...  0.002191  0.005625  0.007540
WBA   0.004019  0.010603  0.008663  ...  0.081536  0.014121  0.005690
WMT   0.014040  0.009843  0.005663  ...  0.014121  0.031133  0.007324
VZ   -0.002546  0.008429  0.010125  ...  0.005690  0.007324  0.022626

[30 rows x 30 columns]

In [68]:
# Set weights and confirm length, i.e. number of stocks
weights = np.array([0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333,	0.0333333])
print(len(weights))
weights



30


array([0.0333333, 0.0333333, 0.0333333, 0.0333333, 0.0333333, 0.0333333,
       0.0333333, 0.0333333, 0.0333333, 0.0333333, 0.0333333, 0.0333333,
       0.0333333, 0.0333333, 0.0333333, 0.0333333, 0.0333333, 0.0333333,
       0.0333333, 0.0333333, 0.0333333, 0.0333333, 0.0333333, 0.0333333,
       0.0333333, 0.0333333, 0.0333333, 0.0333333, 0.0333333, 0.0333333])

In [69]:
sum = 0
for weight in weights:
  sum += weight
print(sum)

0.9999990000000002


In [72]:
# Calculate the portfolio variance
dow_opt_var = np.dot(weights.T, np.dot(dow_opt_cov, weights))
print('Portfolio Variance: ', dow_opt_var)

Portfolio Variance:  0.01514443899809616


In [73]:
# Calculate portfolio volatility
dow_opt_vol = np.sqrt(dow_opt_var)
print('Portfolio Volatility / Standard Deviation: ', dow_opt_vol)

Portfolio Volatility / Standard Deviation:  0.12306274415149436


In [74]:
# Calculate weighted average returns, annualized
dow_opt_port_ret = np.sum(dow_opt_ret.mean()*weights)*252
print('Portfolio Annualized Returns: ', dow_opt_port_ret)

Portfolio Annualized Returns:  0.28422688681390035


In [75]:
# Print Portfolio Summary: Current State
print(" Portfolio Summary")
print("----------------------------")
#print("Assets: ", portfolio_df[:0])
print('Annual Return: ', round(dow_opt_port_ret, 2) * 100, '% / Volatility: ', round(dow_opt_vol, 2) * 100, '% / Variance: ', round(dow_opt_var, 2) * 100, '%')

 Portfolio Summary
----------------------------
Annual Return:  28.000000000000004 % / Volatility:  12.0 % / Variance:  2.0 %


In [77]:
# Portfolio Optimization
# Install portfolio optimization package
!pip install PyPortfolioOpt

     |████████████████████████████████| 60 kB 3.1 MB/s 
     |████████████████████████████████| 2.7 MB 9.0 MB/s 
  Attempting uninstall: cvxpy
    Found existing installation: cvxpy 1.0.31
    Uninstalling cvxpy-1.0.31:
      Successfully uninstalled cvxpy-1.0.31


In [81]:
# import libraries and dependencies
from pypfopt.efficient_frontier import EfficientFrontier
from pypfopt import risk_models
from pypfopt import expected_returns

In [82]:
# Calculate expected annualized returns and sample covariance matrix
mu = expected_returns.mean_historical_return(dow_opt)
S = risk_models.sample_cov(dow_opt)

In [83]:
# Apply modern portfolio theory: max sharpe ratio, efficient ratio, allocate weithts for optimal allocation
ef = EfficientFrontier(mu, S)
weights = ef.max_sharpe()

cleaned_weights = ef.clean_weights()
print(cleaned_weights)
ef.portfolio_performance(verbose=True)

OrderedDict([('AAPL', 0.0), ('AMGN', 0.0), ('AXP', 0.22879), ('BA', 0.0), ('CAT', 0.0), ('CRM', 0.0), ('CSCO', 0.0), ('CVX', 0.0), ('DIS', 0.0), ('DOW', 0.0), ('GS', 0.19552), ('HD', 0.09627), ('HON', 0.0), ('IBM', 0.0), ('INTC', 0.0), ('JNJ', 0.0), ('JPM', 0.0), ('KO', 0.0), ('MCD', 0.0), ('MMM', 0.05392), ('MRK', 0.0), ('MSFT', 0.38287), ('NKE', 0.0), ('PG', 0.0), ('TRV', 0.0), ('UNH', 0.04263), ('V', 0.0), ('WBA', 0.0), ('WMT', 0.0), ('VZ', 0.0)])
Expected annual return: 76.7%
Annual volatility: 14.3%
Sharpe Ratio: 5.23


(0.7668347213661998, 0.14267621564669028, 5.234472459064865)

In [87]:
# Stock Returns
mu

AAPL    0.291723
AMGN    0.205881
AXP     1.036900
BA      0.174433
CAT     0.324956
CRM     0.241683
CSCO    0.551635
CVX     0.387569
DIS    -0.015658
DOW     0.260315
GS      0.892825
HD      0.554945
HON     0.212919
IBM     0.326280
INTC    0.149983
JNJ     0.211117
JPM     0.432982
KO      0.184229
MCD     0.327791
MMM     0.359901
MRK     0.038860
MSFT    0.692587
NKE     0.375472
PG      0.061372
TRV     0.192979
UNH     0.399666
V       0.282271
WBA     0.277021
WMT    -0.036123
VZ     -0.034322
dtype: float64

In [89]:
!pip install pulp

     |████████████████████████████████| 40.6 MB 32 kB/s 


In [92]:
# Portfolio Rebalancing
from pypfopt.discrete_allocation import DiscreteAllocation, get_latest_prices
dow_opt_value = 100000  # variable to hold the amount of money you would like to invest in the optimal portfolio
latest_prices = get_latest_prices(dow_opt)
weights = cleaned_weights
da = DiscreteAllocation(weights, latest_prices, total_portfolio_value = dow_opt_value)
allocation, leftover = da.lp_portfolio()
print('Discrete Allocation: ', allocation)
print('Funds Remaining: $', leftover)

Discrete Allocation:  {'AXP': 133, 'GS': 52, 'HD': 29, 'MMM': 27, 'MSFT': 132, 'UNH': 10}
Funds Remaining: $ 37.42970000000787


Optimal Portfolio for an investment of $100,000
133 shares American Express,
52 shares of Goldman Sachs,
29 shares of Home Depot,
27 shares of 3M,
132 shares of Microsoft,
10 shares of United Healthcare.

Unspent funds: $37.43